File: python/src/modules/examination.py
import random
import secrets
import uuid
from datetime import datetime
from cardinal_sdk import CardinalSdk
from cardinal_sdk.model import DecryptedContact, DecryptedContent, DecryptedPatient, DecryptedService, Identifier, \
	Measure, TimeSeries, DecryptedDocument, DocumentType, DecryptedHealthElement, DecryptedSubContact
from utils import pretty_print_contact


def manage_examination(sdk: CardinalSdk):
	try:
		patient_id = input("Insert the id of a Patient (blank to create a new one): ")
		if len(patient_id) == 0:
			patient = sdk.patient.create_patient_blocking(
				sdk.patient.with_encryption_metadata_blocking(
					DecryptedPatient(
						id=str(uuid.uuid4()),
						first_name="Annabelle",
						last_name="Hall"
					)
				)
			)
		else:
			patient = sdk.patient.get_patient_blocking(patient_id)

		description = input("Examination description: ")
		contact = DecryptedContact(
			id=str(uuid.uuid4()),
			descr=description,
			opening_date=int(datetime.now().strftime("%Y%m%d%H%M%S"))
		)
		contact_with_metadata = sdk.contact.with_encryption_metadata_blocking(contact, patient)
		created_contact = sdk.contact.create_contact_blocking(contact_with_metadata)
		pretty_print_contact(created_contact)

		has_blood_pressure = input("Register blood pressure? [y/N]: ").strip().lower() == "y"
		if has_blood_pressure:
			print("Measuring blood pressure...")
			blood_pressure_service = DecryptedService(
				id=str(uuid.uuid4()),
				label="Blood pressure",
				identifier=[Identifier(system="cardinal", value="bloodPressure")],
				content={
					"en": DecryptedContent(
						measure_value=Measure(
							value=float(random.randint(80, 120)),
							unit="mmHg"
						)
					)
				}
			)
			created_contact.services = [blood_pressure_service]
			contact_with_blood_pressure = sdk.contact.modify_contact_blocking(
				created_contact
			)
		else:
			contact_with_blood_pressure = created_contact

		has_heart_rate = input("Register heart rate? [y/N]: ").strip().lower() == "y"
		if has_heart_rate:
			print("Measuring heart rate...")
			ecg_signal = [round(float(random.uniform(0, 1)), 2) for _ in range(10)]
			heart_rate_service = DecryptedService(
				id=str(uuid.uuid4()),
				identifier=[Identifier(system="cardinal", value="ecg")],
				label="Heart rate",
				content={
					"en": DecryptedContent(
						time_series=TimeSeries(
							samples=[ecg_signal]
						)
					)
				}
			)
			contact_with_blood_pressure.services = contact_with_blood_pressure.services + [heart_rate_service]
			contact_with_ecg = sdk.contact.modify_contact_blocking(contact_with_blood_pressure)
		else:
			contact_with_ecg = contact_with_blood_pressure

		has_x_ray = input("Register x-ray? [y/N]: ").strip().lower() == "y"
		if has_x_ray:
			print("Generating X Ray image...")
			document = DecryptedDocument(
				id=str(uuid.uuid4()),
				document_type=DocumentType.Labresult
			)
			created_contact = sdk.document.create_document_blocking(
				sdk.document.with_encryption_metadata_blocking(document, None)
			)
			x_ray_image = bytearray(secrets.token_bytes(100))
			document_with_attachment = sdk.document.encrypt_and_set_main_attachment_blocking(
				document=created_contact,
				utis=["public.tiff"],
				attachment=x_ray_image
			)
			x_ray_service = DecryptedService(
				id=str(uuid.uuid4()),
				label="X-Ray image",
				identifier=[Identifier(system="cardinal", value="xRay")],
				content={
					"en": DecryptedContent(
						document_id=document_with_attachment.id
					)
				}
			)
			contact_with_ecg.services = contact_with_ecg.services + [x_ray_service]
			contact_with_image = sdk.contact.modify_contact_blocking(contact_with_ecg)
		else:
			contact_with_image = contact_with_ecg

		diagnosis = input("What is the diagnosis?: ")
		health_element = DecryptedHealthElement(
			id=str(uuid.uuid4()),
			descr=diagnosis
		)
		created_diagnosis = sdk.health_element.create_health_element_blocking(
			sdk.health_element.with_encryption_metadata_blocking(health_element, patient)
		)
		contact_with_image.sub_contacts = [
			DecryptedSubContact(
				descr="Diagnosis",
				health_element_id=created_diagnosis.id
			)
		]
		contact_with_diagnosis = sdk.contact.modify_contact_blocking(contact_with_image)

		close = input("Close contact? [y/N]: ").strip().lower() == "y"
		if close:
			contact_with_diagnosis.closing_date = int(datetime.now().strftime("%Y%m%d%H%M%S"))
			final_contact = sdk.contact.modify_contact_blocking(contact_with_diagnosis)
		else:
			final_contact = contact_with_diagnosis

		pretty_print_contact(final_contact)
	except Exception as e:
		print(f"Something went wrong: {e}")

File: kotlin/src/main/kotlin/com/cardinal/modules/Examination.kt
package com.cardinal.modules

import com.icure.cardinal.sdk.model.DecryptedContact
import com.icure.cardinal.sdk.model.DecryptedDocument
import com.icure.cardinal.sdk.model.DecryptedHealthElement
import com.icure.cardinal.sdk.model.DecryptedPatient
import com.icure.cardinal.sdk.model.base.Identifier
import com.icure.cardinal.sdk.model.embed.DecryptedContent
import com.icure.cardinal.sdk.model.embed.DecryptedService
import com.icure.cardinal.sdk.model.embed.DecryptedSubContact
import com.icure.cardinal.sdk.model.embed.DocumentType
import com.icure.cardinal.sdk.model.embed.Measure
import com.icure.cardinal.sdk.model.embed.TimeSeries
import com.cardinal.utils.prettyPrint
import com.icure.cardinal.sdk.CardinalSdk
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.UUID
import kotlin.random.Random

suspend fun createConsultation(sdk: CardinalSdk) {
	try {
		print("Insert the id of a Patient (blank to create a new one): ")
		val patientId = readlnOrNull()
		val patient = if(patientId.isNullOrBlank()) {
			sdk.patient.createPatient(
				DecryptedPatient(
					id = UUID.randomUUID().toString(),
					firstName = "Annabelle",
					lastName = "Hall",
				).let { sdk.patient.withEncryptionMetadata(it) }
			)
		} else {
			sdk.patient.getPatient(patientId)
		}

		val formatter = DateTimeFormatter.ofPattern("yyyyMMddHHmmss")
		print("Examination description: ")
		val description = readln()
		val contact = DecryptedContact(
			id = UUID.randomUUID().toString(),
			descr = description,
			openingDate = LocalDateTime.now().format(formatter).toLong()
		)
		val contactWithMetadata = sdk.contact.withEncryptionMetadata(contact, patient)
		val createdContact = sdk.contact.createContact(contactWithMetadata)
		prettyPrint(createdContact)

		print("Register blood pressure? [y/N]: ")
		val hasBloodPressure = readln().trim().lowercase() == "y"
		val contactWithBloodPressure = if (hasBloodPressure) {
			println("Measuring blood pressure...")
			val bloodPressureService = DecryptedService(
				id = UUID.randomUUID().toString(),
				label = "Blood pressure",
				identifier = listOf(Identifier(system = "icure", value = "bloodPressure")),
				content = mapOf(
					"en" to DecryptedContent(
						measureValue = Measure(
							value = Random.nextInt(80, 120).toDouble(),
							unit = "mmHg"
						)
					)
				)
			)
			sdk.contact.modifyContact(
				createdContact.copy(
					services = setOf(bloodPressureService)
				)
			)
		} else createdContact

		print("Register heart rate? [y/N]: ")
		val hasHeartRate = readln().trim().lowercase() == "y"
		val contactWithECG = if(hasHeartRate) {
			println("Measuring heart rate...")
			val ecgSignal = List(10) { Random.nextInt(0, 100) / 100.0 }
			val heartRateService = DecryptedService(
				id = UUID.randomUUID().toString(),
				identifier = listOf(Identifier(system = "icure", value = "ecg")),
				label = "Heart rate",
				content = mapOf(
					"en" to DecryptedContent(
						timeSeries = TimeSeries(
							samples = listOf(ecgSignal)
						)
					)
				)
			)
			sdk.contact.modifyContact(
				contactWithBloodPressure.copy(
					services = contactWithBloodPressure.services + heartRateService
				)
			)
		} else contactWithBloodPressure

		print("Register x-ray? [y/N]: ")
		val hasXRay = readln().trim().lowercase() == "y"
		val contactWithImage = if(hasXRay) {
			println("Generating X Ray image...")
			val document = DecryptedDocument(
				id = UUID.randomUUID().toString(),
				documentType = DocumentType.Labresult
			)
			val createdDocument = sdk.document.createDocument(
				sdk.document.withEncryptionMetadata(document, null)
			)
			val xRayImage = Random.nextBytes(100)
			val documentWithAttachment = sdk.document.encryptAndSetMainAttachment(
				document = createdDocument,
				utis = listOf("public.tiff"),
				attachment = xRayImage
			)
			val xRayService = DecryptedService(
				id = UUID.randomUUID().toString(),
				label = "X-Ray image",
				identifier = listOf(Identifier(system = "icure", value = "xRay")),
				content = mapOf(
					"en" to DecryptedContent(
						documentId = documentWithAttachment.id
					)
				)
			)
			sdk.contact.modifyContact(
				contactWithECG.copy(
					services = contactWithECG.services + xRayService,
				)
			)
		} else contactWithECG

		print("What is the diagnosis?: ")
		val diagnosis = readln().trim()
		val healthElement = DecryptedHealthElement(
			id = UUID.randomUUID().toString(),
			descr = diagnosis
		)
		val createdDiagnosis = sdk.healthElement.createHealthElement(
			sdk.healthElement.withEncryptionMetadata(healthElement, patient)
		)
		val contactWithDiagnosis = sdk.contact.modifyContact(
			contactWithImage.copy(
				subContacts = setOf(DecryptedSubContact(
					descr = "Diagnosis",
					healthElementId = createdDiagnosis.id
				))
			)
		)

		print("Close contact? [y/N]: ")
		val close = readln().trim() == "y"

		val finalContact = if (close) {
			sdk.contact.modifyContact(contactWithDiagnosis.copy(
				closingDate = LocalDateTime.now().format(formatter).toLong()
			))
		} else contactWithImage

		prettyPrint(finalContact)

	} catch (e: Exception) {
		println("Something went wrong: ${e.message}")
	}
}